File: 242-valid-anagram/valid-anagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    if (s.length !== t.length) return false

    const map = new Map()

    for (const char of s) {
        map.set(char, (map.get(char) || 0) + 1)
    }

    for (const char of t) {
        if (!map.has(char)) {
            return false
        } else {
            map.set(char, (map.get(char) || 0) - 1)
            if (map.get(char) === 0) {
                map.delete(char)
            }
        }
    }

    return map.size === 0
};

File: 242-valid-anagram/valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t): return False

        count = {}

        for char in s:
            count[char] = count.get(char, 0) + 1

        for char in t:
            if char not in count or count[char] == 0:
                return False
            count[char] -= 1

        return True